src/self-managed-debug/src/main.rs

```diff
@@ -0,0 +1,451 @@
+            K8sResourceType::Service => "services",
+            K8sResourceType::Deployment => "deployments",
+            K8sResourceType::StatefulSet => "statefulsets",
+            K8sResourceType::ReplicaSet => "replicasets",
+            K8sResourceType::NetworkPolicy => "networkpolicies",
+            K8sResourceType::Event => "events",
+            K8sResourceType::Node => "nodes",
+        };
+        write!(f, "{}", s)
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let args: Args = cli::parse_args(CliConfig {
+        env_prefix: Some("SELF_MANAGED_DEBUG_"),
+        enable_version_flag: true,
+    });
+
+    let start_time = Utc::now();
+
+    let context = Context { start_time, args };
+
+    if let Err(err) = run(context).await {
+        eprintln!(
+            "self-managed-debug: fatal: {}\nbacktrace: {}",
+            err.display_with_causes(),
+            err.backtrace()
+        );
+        process::exit(1);
+    }
+}
+
+async fn run(context: Context) -> Result<(), anyhow::Error> {
+    let mut describe_handles = Vec::new();
+
+    let mut spawn_kubectl_describe_process =
+        |resource_type: &K8sResourceType, namespace: Option<String>| {
+            let context = context.clone();
+            let resource_type = resource_type.clone();
+            let namespace = namespace.clone();
+
+            // Spawn a new task for each describe operation
+            let handle = mz_ore::task::spawn(|| "dump-kubectl-describe", async move {
+                if let Err(e) =
+                    dump_kubectl_describe(&context, &resource_type, namespace.as_ref()).await
+                {
+                    eprintln!(
+                        "Failed to dump kubectl describe for {}: {}",
+                        resource_type, e
+                    );
+                }
+            });
+            describe_handles.push(handle);
+        };
+
+    for resource_type in [
+        K8sResourceType::Pod,
+        K8sResourceType::Service,
+        K8sResourceType::Deployment,
+        K8sResourceType::StatefulSet,
+        K8sResourceType::ReplicaSet,
+        K8sResourceType::NetworkPolicy,
+        K8sResourceType::Event,
+    ] {
+        for namespace in context.args.k8s_namespaces.clone() {
+            spawn_kubectl_describe_process(&resource_type, Some(namespace));
+        }
+    }
+
+    spawn_kubectl_describe_process(&K8sResourceType::Node, None);
+
+    match create_k8s_client(context.args.k8s_context.clone()).await {
+        Ok(client) => {
+            for namespace in context.args.k8s_namespaces.clone() {
+                if let Err(e) = dump_k8s_pod_logs(&context, client.clone(), &namespace).await {
+                    eprintln!(
+                        "Failed to write k8s pod logs for namespace {}: {}",
+                        namespace, e
+                    );
+                }
+
+                let events_api = Api::<Event>::namespaced(client.clone(), &namespace);
```

Comment:
I initially wanted to enumerate each API resource type and loop over them instead of going one by one. Maybe even map them to K8sResource type. But I figured it was more code than and more complex than worth. Lemme know if it's easier than I think though!